File: android/src/main/kotlin/com/easemob/flutter_callkit/EasemobFlutterCallkitPlugin.kt
package com.easemob.flutter_callkit

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.hyphenate.EMCallBack
import com.hyphenate.chat.EMClient
import com.hyphenate.chat.EMOptions
import com.hyphenate.callkit.CallKitClient
import com.hyphenate.callkit.CallKitConfig
import com.hyphenate.callkit.bean.CallEndReason
import com.hyphenate.callkit.bean.CallInfo
import com.hyphenate.callkit.bean.CallKitGroupInfo
import com.hyphenate.callkit.bean.CallKitUserInfo
import com.hyphenate.callkit.bean.CallType
import com.hyphenate.callkit.interfaces.CallInfoProvider
import com.hyphenate.callkit.interfaces.CallKitListener
import com.hyphenate.chat.EMUserInfo
import com.hyphenate.EMValueCallBack
import org.json.JSONObject
import org.json.JSONException

/** EasemobFlutterCallkitPlugin */
class EasemobFlutterCallkitPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "easemob_flutter_callkit")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "easemob_flutter_callkit/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initIM" -> {
                val appKey = call.argument<String>("appKey")
                val autoLogin = call.argument<Boolean>("autoLogin") ?: false
                if (appKey == null) {
                    result.error("INVALID_ARGUMENT", "appKey is required", null)
                    return
                }
                val options = EMOptions().apply {
                    setAppKey(appKey)
                    setAutoLogin(autoLogin)
                }
                context?.let { ctx ->
                    EMClient.getInstance().init(ctx, options)
                    result.success(true)
                } ?: result.error("NO_CONTEXT", "Context is null", null)
            }
            "initCallKit" -> {
                try {
                    EMClient.getInstance().chatManager()
                } catch (e: Exception) {
                    result.error("IM_NOT_INITIALIZED", "Please call initIM() before initCallKit()", null)
                    return
                }
                val configMap = call.argument<Map<String, Any?>>("config")
                val config = CallKitConfig().apply {
                    configMap?.get("callTimeout")?.let {
                        callTimeout = (it as Number).toInt()
                    }
                    configMap?.get("incomingRingFile")?.let {
                        incomingRingFile = it as String
                    }
                    configMap?.get("outgoingRingFile")?.let {
                        outgoingRingFile = it as String
                    }
                    configMap?.get("dingRingFile")?.let {
                        dingRingFile = it as String
                    }
                }
                context?.let { ctx ->
                    try {
                        CallKitClient.init(ctx, config)
                        CallKitClient.callInfoProvider = DefaultCallInfoProvider()
                        setupCallKitListener()
                        result.success(true)
                    } catch (e: Exception) {
                        result.error("INIT_CALLKIT_ERROR", e.message, null)
                    }
                } ?: result.error("NO_CONTEXT", "Context is null", null)
            }
            "login" -> {
                val username = call.argument<String>("username")
                val token = call.argument<String>("token")
                if (username == null || token == null) {
                    result.error("INVALID_ARGUMENT", "username and token are required", null)
                    return
                }
                EMClient.getInstance().loginWithToken(username, token, object : EMCallBack {
                    override fun onSuccess() {
                        result.success(true)
                    }
                    override fun onError(code: Int, error: String?) {
                        result.error("LOGIN_ERROR", error ?: "Unknown error", code)
                    }
                })
            }
            "loginWithPassword" -> {
                val username = call.argument<String>("username")
                val password = call.argument<String>("password")
                if (username == null || password == null) {
                    result.error("INVALID_ARGUMENT", "username and password are required", null)
                    return
                }
                EMClient.getInstance().login(username, password, object : EMCallBack {
                    override fun onSuccess() {
                        result.success(true)
                    }
                    override fun onError(code: Int, error: String?) {
                        result.error("LOGIN_ERROR", error ?: "Unknown error", code)
                    }
                })
            }
            "logout" -> {
                EMClient.getInstance().logout(true, object : EMCallBack {
                    override fun onSuccess() {
                        result.success(true)
                    }
                    override fun onError(code: Int, error: String?) {
                        result.error("LOGOUT_ERROR", error ?: "Unknown error", code)
                    }
                })
            }
            "startSingleCall" -> {
                val callType = call.argument<Int>("callType") ?: 0
                val userId = call.argument<String>("userId")
                val ext = call.argument<String>("ext")
                if (userId == null) {
                    result.error("INVALID_ARGUMENT", "userId is required", null)
                    return
                }
                val type = if (callType == 0) CallType.SINGLE_VOICE_CALL else CallType.SINGLE_VIDEO_CALL
                val extObj = ext?.let { parseExtToJSONObject(it) }
                CallKitClient.startSingleCall(type, userId, extObj)
                result.success(true)
            }
            "startGroupCall" -> {
                val groupId = call.argument<String>("groupId")
                val ext = call.argument<String>("ext")
                if (groupId == null) {
                    result.error("INVALID_ARGUMENT", "groupId is required", null)
                    return
                }
                val extObj = ext?.let { parseExtToJSONObject(it) }
                CallKitClient.startGroupCall(groupId, extObj)
                result.success(true)
            }
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun parseExtToMap(ext: String): Map<String, Any>? {
        return try {
            val json = JSONObject(ext)
            val map = mutableMapOf<String, Any>()
            val keys = json.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                map[key] = json.get(key)
            }
            map
        } catch (e: JSONException) {
            null
        }
    }

    private fun parseExtToJSONObject(ext: String): JSONObject? {
        return try {
            JSONObject(ext)
        } catch (e: JSONException) {
            null
        }
    }

    private fun setupCallKitListener() {
        CallKitClient.callKitListener = object : CallKitListener {
            override fun onReceivedCall(userId: String, callType: CallType, ext: JSONObject?) {
                val event = mapOf(
                    "event" to "onReceivedCall",
                    "callType" to callType.code,
                    "userId" to userId,
                    "ext" to ext?.toString()
                )
                sendEvent(event)
            }

            override fun onEndCallWithReason(reason: CallEndReason, callInfo: CallInfo?) {
                val event = mapOf(
                    "event" to "onEndCallWithReason",
                    "callType" to (callInfo?.callKitType?.code ?: -1),
                    "channelName" to (callInfo?.channelName ?: ""),
                    "reason" to reason.code,
                    "time" to (callInfo?.callTime ?: 0L)
                )
                sendEvent(event)
            }

            override fun onCallError(errorType: CallKitClient.CallErrorType, errorCode: Int, description: String?) {
                val event = mapOf(
                    "event" to "onCallError",
                    "code" to errorCode,
                    "description" to description
                )
                sendEvent(event)
            }

            override fun onRemoteUserJoined(userId: String, callType: CallType, channelName: String) {
                val event = mapOf(
                    "event" to "onRemoteUserJoined",
                    "channelName" to channelName,
                    "userId" to userId
                )
                sendEvent(event)
            }

            override fun onRemoteUserLeft(userId: String, callType: CallType, channelName: String) {
                val event = mapOf(
                    "event" to "onRemoteUserLeft",
                    "channelName" to channelName,
                    "userId" to userId
                )
                sendEvent(event)
            }

            override fun onRtcEngineCreated(engine: io.agora.rtc2.RtcEngine) {
                // Not forwarded to Dart by default
            }
        }
    }

    private fun sendEvent(event: Map<String, Any?>) {
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            eventSink?.success(event)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}

class DefaultCallInfoProvider : CallInfoProvider {
    override fun asyncFetchUsers(
        userIds: List<String>,
        onValueSuccess: com.hyphenate.callkit.interfaces.OnValueSuccess<List<CallKitUserInfo>>
    ) {
        EMClient.getInstance().userInfoManager().fetchUserInfoByAttribute(
            userIds.toTypedArray(),
            arrayOf(EMUserInfo.EMUserInfoType.NICKNAME, EMUserInfo.EMUserInfoType.AVATAR_URL),
            object : EMValueCallBack<Map<String, EMUserInfo>> {
                override fun onSuccess(value: Map<String, EMUserInfo>?) {
                    val userInfos = mutableListOf<CallKitUserInfo>()
                    value?.forEach { (userId, info) ->
                        userInfos.add(CallKitUserInfo(
                            userId = userId,
                            nickName = info.nickname,
                            avatar = info.avatarUrl
                        ))
                    }
                    onValueSuccess(userInfos)
                }

                override fun onError(error: Int, errorMsg: String?) {
                    onValueSuccess(emptyList())
                }
            }
        )
    }

    override fun asyncFetchGroupInfo(
        groupId: String,
        onValueSuccess: com.hyphenate.callkit.interfaces.OnValueSuccess<CallKitGroupInfo>
    ) {
        val groups = EMClient.getInstance().groupManager().allGroups
        val group = groups?.find { it.groupId == groupId }
        val groupInfo = CallKitGroupInfo(
            groupID = groupId,
            groupName = group?.groupName,
            groupAvatar = group?.getGroupAvatar()
        )
        onValueSuccess(groupInfo)
    }
}

File: ios/Classes/EasemobFlutterCallkitPlugin.swift
import Flutter
import UIKit
import EaseCallUIKit
import HyphenateChat
import AgoraRtcKit

public class EasemobFlutterCallkitPlugin: NSObject, FlutterPlugin {
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "easemob_flutter_callkit", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "easemob_flutter_callkit/events", binaryMessenger: registrar.messenger())

        let instance = EasemobFlutterCallkitPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initIM":
            guard let args = call.arguments as? [String: Any],
                  let appKey = args["appKey"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "appKey is required", details: nil))
                return
            }
            let autoLogin = args["autoLogin"] as? Bool ?? false
            let options = EMOptions(appkey: appKey)
            options.isAutoLogin = autoLogin
            EMClient.shared().initializeSDK(with: options)
            result(true)

        case "initCallKit":
            let config = EaseCallUIKit.CallKitConfig()
            config.enableVOIP = true
            config.enablePIPOn1V1VideoScene = true
            if let args = call.arguments as? [String: Any],
               let configMap = args["config"] as? [String: Any] {
                if let timeout = configMap["callTimeout"] as? Int {
                    // CallKitConfig doesn't expose direct timeout setter in public API, keep for Dart layer
                }
            }
            CallKitManager.shared.setup(config)
            CallKitManager.shared.profileProvider = self
            setupListeners()
            result(true)

        case "login":
            guard let args = call.arguments as? [String: Any],
                  let username = args["username"] as? String,
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "username and token are required", details: nil))
                return
            }
            EMClient.shared().login(withUsername: username, token: token) { username, error in
                if let error = error {
                    result(FlutterError(code: "LOGIN_ERROR", message: error.errorDescription, details: error.code))
                } else {
                    result(true)
                }
            }

        case "loginWithPassword":
            guard let args = call.arguments as? [String: Any],
                  let username = args["username"] as? String,
                  let password = args["password"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "username and password are required", details: nil))
                return
            }
            EMClient.shared().login(withUsername: username, password: password) { username, error in
                if let error = error {
                    result(FlutterError(code: "LOGIN_ERROR", message: error.errorDescription, details: error.code))
                } else {
                    result(true)
                }
            }

        case "logout":
            EMClient.shared().logout(true) { error in
                if let error = error {
                    result(FlutterError(code: "LOGOUT_ERROR", message: error.errorDescription, details: error.code))
                } else {
                    result(true)
                }
            }

        case "startSingleCall":
            guard let args = call.arguments as? [String: Any],
                  let callTypeInt = args["callType"] as? Int,
                  let userId = args["userId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "callType and userId are required", details: nil))
                return
            }
            let type: CallType = (callTypeInt == 0) ? .singleAudio : .singleVideo
            CallKitManager.shared.call(with: userId, type: type)
            result(true)

        case "startGroupCall":
            guard let args = call.arguments as? [String: Any],
                  let groupId = args["groupId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "groupId is required", details: nil))
                return
            }
            CallKitManager.shared.groupCall(groupId: groupId)
            result(true)

        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func setupListeners() {
        CallKitManager.shared.addListener(self)
    }

    private func sendEvent(_ event: [String: Any?]) {
        eventSink?(event)
    }

    private func callInfoToDict(_ info: CallInfo?) -> [String: Any?]? {
        guard let info = info else { return nil }
        return [
            "callId": info.callId,
            "callerId": info.callerId,
            "calleeId": info.calleeId,
            "channelName": info.channelName,
            "type": info.type.rawValue,
            "duration": info.duration,
            "groupId": info.groupId,
            "extensionInfo": info.extensionInfo,
        ]
    }
}

extension EasemobFlutterCallkitPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

extension EasemobFlutterCallkitPlugin: CallServiceListener {
    public func onReceivedCall(callType: CallType, userId: String, extensionInfo: [String : Any]?) {
        sendEvent([
            "event": "onReceivedCall",
            "callType": callType.rawValue,
            "userId": userId,
            "ext": extensionInfo
        ])
    }
    public func didOccurError(error: CallError) {
        sendEvent([
            "event": "onCallError",
            "code": error.error.errorCode,
            "description": error.errorMessage
        ])
    }

    public func didUpdateCallEndReason(reason: CallEndReason, info: CallInfo) {
        sendEvent([
            "event": "onEndCallWithReason",
            "reason": reason.rawValue,
            "info": callInfoToDict(info)
        ])
    }

    public func remoteUserDidJoined(userId: String, channelName: String, type: CallType) {
        sendEvent([
            "event": "onRemoteUserJoined",
            "userId": userId,
            "channelName": channelName,
            "callType": type.rawValue
        ])
    }

    public func remoteUserDidLeft(userId: String, channelName: String, type: CallType) {
        sendEvent([
            "event": "onRemoteUserLeft",
            "userId": userId,
            "channelName": channelName,
            "callType": type.rawValue
        ])
    }

    public func onRtcEngineCreated(engine: AgoraRtcEngineKit) {
        // Allow users to configure engine through Dart layer if needed
    }
}

extension EasemobFlutterCallkitPlugin: CallUserProfileProvider {
    public func fetchUserProfiles(profileIds: [String]) async -> [any CallProfileProtocol] {
        var resultProfiles: [any CallProfileProtocol] = []
        var unknownIds: [String] = []
        for profileId in profileIds {
            if let profile = CallKitManager.shared.usersCache[profileId] {
                resultProfiles.append(profile)
            } else {
                unknownIds.append(profileId)
            }
        }
        guard !unknownIds.isEmpty else {
            return resultProfiles
        }
        let userInfos = await withCheckedContinuation { (continuation: CheckedContinuation<[String: EMUserInfo]?, Never>) in
            EMClient.shared().userInfoManager?.fetchUserInfo(byId: unknownIds) { infos, _ in
                continuation.resume(returning: infos)
            }
        }
        if let infoMap = userInfos {
            for (userId, info) in infoMap {
                let profile = CallUserProfile()
                profile.id = userId
                profile.nickname = info.nickname ?? ""
                profile.avatarURL = info.avatarUrl ?? ""
                resultProfiles.append(profile)
            }
        }
        return resultProfiles
    }

    public func fetchGroupProfiles(profileIds: [String]) async -> [any CallProfileProtocol] {
        var resultProfiles: [any CallProfileProtocol] = []
        let groups = EMClient.shared().groupManager?.getJoinedGroups() ?? []
        for groupId in profileIds {
            if let group = groups.first(where: { $0.groupId == groupId }) {
                let profile = CallUserProfile()
                profile.id = groupId
                profile.nickname = group.groupName
                profile.avatarURL = group.groupAvatar ?? ""
                resultProfiles.append(profile)
            }
        }
        return resultProfiles
    }
}
